internal/tracing/tracing.go

```diff
@@ -0,0 +1,66 @@
+package tracing
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+)
+
+type tpCtxKey struct{}
+
+// Context creates a context with a TracerProvider added to it which the TracerProvider and Tracer functions can
+// retrieve. It attempts to create a TracerProvider which exports traces over OTLP+HTTP to the URL defined in the
+// environment. If that's not possible, it returns the supplied context unnmodified and an error.
+// If both OTEL_EXPORTER_OTLP_ENDPOINT and OTEL_EXPORTER_OTLP_TRACES_ENDPOINT are unset, instead of returning a broken
+// exporter (otel default behavior), this function returns none.
+func Context(ctx context.Context) (context.Context, error) {
+	if os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT") == "" && os.Getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT") == "" {
+		// Otel is not configured, do not enable tracing.
+		return ctx, nil
+	}
```

Comment:
Unfortunately checking for env var emptyness is hard to avoid. The otel library has a semi-complex logic of preferring one or the other that I'd like to preserve intact, as such is **the otel way**. Using env vars to configure tracing is also the otel way.

Unfortunately for us, if they are unset the otel libraries will not return an error, but a broken tracer.

Again I'd rather not try to fix fundamental issues with the otel libraries, but stick to what everybody is doing and expecting.

Reply:
I've moved the env-checking code and fallback-to-noop to main, LMK if this looks better.

Reply:
Continuing discussion elsewhere.